### PR TITLE
WIP: Rectified/Insta Flow

### LINKED ISSUE #5256: Add Insta-Flow Training support
**Is your feature request related to a problem? Please describe.**
For hacktoberfest, I wanted to add in the Rectified Flow training objective which is what the instaflow paper, which can generate images in one forward pass, used to finetune stable diffusion.

**Describe the solution you'd like**
I'll make a small pr and argument to examples/text_to_image based on code from [here](https://github.com/gnobitab/RectifiedFlow/blob/main/ImageGeneration/losses.py) to examples/text_to_image loss. I'm thinking of testing it out with mnist to see if it works.

**Additional context**
The links to the papers:
[instaflow](https://arxiv.org/abs/2309.06380)
[rectified flow](https://arxiv.org/abs/2209.03003)

### PR DESCRIPTION
**Pull request by isamu-isozaki**
# What does this PR do?
Introduces Rectified/Instaflow as an example for diffusers. This is a WIP PR discussed in [this issue](https://github.com/huggingface/diffusers/issues/5256). The main part should be implemented. For testing, I'm thinking of using [bk-sdm tiny](https://huggingface.co/nota-ai/bk-sdm-tiny) for the model and for the dataset [lambdalabs/pokemon-blip-captions](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions). Let me know if anyone knows a cheaper sd model! I'm using sd so that I don't deviate too much from train_text_to_image.py

Once all the parts are tested, I think this pr is done! As a todo list, the tasks I can think of are

- [ ] Test Dataset is created correctly
- [ ] Test Rectified Flow works
- [ ] Test Distilling works

Fixes #5256 (issue)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case. 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by isamu-isozaki** [comment]
Creating a dataset seems to work. The only part missing is adding LPIPs loss but that can be a separate PR once we have this merged. Currently, I'm thinking the bk-sdm-tiny might not be a great demonstration so I'm trying nota-ai/bk-sdm-small and OFA-Sys/small-stable-diffusion-v0. Let me know if anyone knows a smaller text-to-image model(ideally like stable diffusion)!

**Comment by patrickvonplaten** [comment]
Ping again here @DN6

**Comment by isamu-isozaki** [comment]
Ok! I'm thinking bk-sdm might be small enough. I have a colab script [here](https://colab.research.google.com/drive/13GmggtFLnVj55i2XnVp-E_-tLgv2ThSm#scrollTo=9QpxgPmqXZQd) which I plan to use to test if rectified flow properly trains. Will post results here as they come in. On 6gb vram I wasn't able to train it

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽

**Comment by DN6** [comment]
@isamu-isozaki Could you run `make style && make quality` on your PR. I think once you can verify that rectified flow  trains properly, we can merge this.

**Comment by isamu-isozaki** [comment]
@DN6 sounds good! Will try testing a minimal example asap

**Comment by danieltanhx** [comment]
possible to create train_lora_rectified_instaflow.py soon? Computationally/in terms of storage, it might be too much for others to test it out on the full model.

**Comment by isamu-isozaki** [comment]
Just 3000 steps but it seems like unet is struggling a bit in converting from ddpm objective to rectified flow objective. As a reference, for InstaFlow they did 100k steps which I'm hoping to avoid for this minimal example.
imageData

And @danieltanhx tnx for the idea! I was having trouble with compute for testing so might try this.

**Comment by isamu-isozaki** [comment]
I'll resume this test with lower res/smaller model to see if it works. I might switch to non-sd models if smaller

**Comment by isamu-isozaki** [comment]
@danieltanhx actually let me see if lora works for this haha

**Comment by isamu-isozaki** [comment]
Okay for lora I can run it pretty fast in Colab. So far, the loss is going down but will see if this is enough once I am able to use colab again

loss

**Comment by adhikjoshi** [comment]
https://github.com/gnobitab/InstaFlow
